### ISSUE TITLE
Control plane operator design doc

### ISSUE THREAD

**Issue by luomai** [lsds/KungFu#41]
# Design goals

1. Computing monitoring metrics for the learning plane. The first monitoring metric is: gradient noise scale proposed by OpenAI. 
2. Applying changes to hyper-parameters on distributed devices. This can be achieved using the Chi approach where we maintain a distributed asynchronous control barrier for workers.

# Dependency graph of KungFu dataflows
Note: (A, B) means operator B has a dependency on operator A.

A TensorFlow dataflow:

```
(forward, loss)
(loss, gradients)
(gradients, reduced_gradients)
(reduced_gradients, optimizer)
(optimizer, forward)
.....
```

A controllable TensorFlow dataflow

```
(forward, loss)
(loss, gradients)
(gradients, reduced_gradients)
(gradients, gradient_noise)
(reduced_gradients, optimizer)
(gradient_noise, controller)
(optimizer, controller)
(controller, forward)
......
```
